src/gs-update-monitor.c

```diff
@@ -500,6 +459,72 @@ schedule_update (GsUpdateMonitor *monitor,
 				       schedule_entry_scheduled_cb, helper);
 }
 
+static void
+schedule_updates_real (GsUpdateMonitor *monitor,
+		       GsAppList *apps_to_update)
+{
+	/* we have to use an updates counter to be able to know when all
+	 * updates have been scheduled */
+	monitor->num_scheduled_updates = gs_app_list_length (apps_to_update);
+	for (guint i = 0; i < monitor->num_scheduled_updates; i++) {
+		GsApp *app = gs_app_list_index (apps_to_update, i);
+		schedule_update (monitor, app,
+				 monitor->scheduled_updates_cancellable);
+	}
+}
+
+static void
+monitor_clear_scheduler (GsUpdateMonitor *monitor)
+{
+	g_clear_object (&monitor->scheduler);
+}
+
+static void
+scheduler_ready_cb (GObject *source_object,
+		    GAsyncResult *result,
+		    gpointer data)
+{
+	MwscScheduler *scheduler;
+	g_autoptr(GError) local_error = NULL;
+	g_autoptr(SchedulerHelper) helper = (SchedulerHelper *) data;
+	GsUpdateMonitor *monitor = helper->monitor;
+
+	scheduler = mwsc_scheduler_new_finish (result, &local_error);
+
+	if (scheduler == NULL) {
+		g_warning ("Error getting Mogwai Scheduler: %s", local_error->message);
+		return;
+	}
+
+	g_signal_connect_object (scheduler, "invalidated",
+				 (GCallback) monitor_clear_scheduler,
+				 monitor,
+				 G_CONNECT_SWAPPED);
+	monitor->scheduler = scheduler;
+
+	schedule_updates_real (monitor, helper->apps_to_update);
+}
+
+static void
+schedule_updates (GsUpdateMonitor *monitor,
+		  GsAppList *apps_to_update)
+{
+	/* if we don't have a valid scheduler yet, create it asynchronously and
+	 * delegate the updates scheduling to its callback */
+	if (monitor->scheduler == NULL) {
+		SchedulerHelper *helper = g_new0 (SchedulerHelper, 1);
+		helper->monitor = monitor;
+		helper->apps_to_update = apps_to_update;
```

Comment:
Nitpick: Use `g_steal_pointer (&apps_to_update)` here to make it obvious you’re transferring ownership to the `SchedulerHelper`?

Reply:
I pondered that and decided not to do it but yeah, maybe it's more future proof.

Reply:
(meaning I will do it :smile: )

Reply:
I thought you were actually referring to the `helper`, sorry. I will add something to both vars in any case. Stay tuned :)

Reply:
:+1: